### PR TITLE
Rust: Improve lines-of-code counts.

### PR DESCRIPTION
**Pull request by geoffw0**
Improve lines-of-code counts:
- remove `Module` from the count, which was spuriously adding +1 LOC in `main.rs`.
- add back calls to `println!` in the tests, since these should be deterministically (if not correctly) extracted now.
- coincidentally the total LOC counted changes after the first commit, but changes back exactly after the second.  These changes are not doing nothing!

@aibaars  - also note that we have a spuriously located `MacroRules` (on `my_macro.rs` line 1) and `Struct` (on `my_struct.rs` line 2), which I discovered while debugging this query.  I haven't done anything about them, I believe the right fix would be in the extractor.

Obviously types are still not extracted, which is also affecting the counts.

### CODE REVIEW

File: rust/ql/lib/codeql/files/FileSystem.qll [hunk 1]
```diff
@@ -44,7 +45,9 @@ class File extends Container, Impl::File {
     result =
       count(int line |
         exists(AstNode node, Location loc |
-          not node instanceof SourceFile and loc = node.getLocation()
+          not node instanceof Module and
```
Comment (aibaars) [inline rust/ql/lib/codeql/files/FileSystem.qll:48]: Please revert this change. The module isn't the problem. The problem are the `/* */` comments which syntactically belong to the `mod`, `struct`, `macrorule`, etc. that follows them. Just verify for yourself using the AST viewer.

**Review by aibaars — CHANGES_REQUESTED** [review]

File: rust/ql/lib/codeql/files/FileSystem.qll (hunk 1 above)
Comment (geoffw0) [reply rust/ql/lib/codeql/files/FileSystem.qll:48]: Hmm, I see what you mean now, `Module` is a module decl not a module.  I'm going to have to get used to that.

Comment (aibaars) [reply rust/ql/lib/codeql/files/FileSystem.qll]: I think the AST viewer works fine on test databases these days. I just built a database running `codeql test run --keep-databases`, loaded it in vscode, ran `from File f select f`, click on a file, and right-click "view AST" in that file.

Comment (geoffw0) [reply rust/ql/lib/codeql/files/FileSystem.qll]: That method for tests is a bit of a fiddle unfortunately, but good to know the AST viewer now works for Rust.  I had investigated using an adhoc query to print all the `AstNode.toString()`s associated with each line of each file, which worked - though I didn't get a hierarchy.

Anyway I've created an issue for the locations problems.

**Comment by aibaars** [comment]
In general using `start` and `end` lines of AstNodes to count lines will never work 100%. Currently we're overcounting a bit for items with attached comments. However, we may also change the location of top-level items to be the span of their names only. That gives a much better user-experience (so we highlightithe name of the function instead of the entire function body). Another tricky thing to deal with are multi-line string literals. The only reliable way to calculate LOC is to look at token, which we do not extract (though we might at some point).

A pretty simple way to deal with this problem is to ignore the start line for top-level items. That would under count in cases like 
```rust
fn
my_function() {...}
```
but that would be extremely rare.

**Review by aibaars — APPROVED** [review]
Changes look good to me. However, you could implement special handling of top-level items to avoid counting doc comments as a line.

**Comment by geoffw0** [comment]
Merged main, resolved conflicts and excluded top-level AST elements from the count (for now).

> In general using start and end lines of AstNodes to count lines will never work 100%.

True, but notably this has proven to be a good enough approximation in other languages.  Multi-line string literals should be pretty rare, and we might even add special logic for them if it turns out to be a thing that bothers us.

File: rust/ql/lib/codeql/files/FileSystem.qll [hunk 2]
```diff
@@ -47,6 +48,7 @@ class File extends Container, Impl::File {
         exists(AstNode node, Location loc |
           not node instanceof Comment and
           not node instanceof SourceFile and
+          not getImmediateParent(node) instanceof SourceFile and // ignore top-level elements for now as we're getting their locations wrong when a comment is attached
```
Comment (aibaars) [inline rust/ql/lib/codeql/files/FileSystem.qll:51]: Perhaps just `not node instanceof Item` . That should also deal with `Item`s that are nested in functions.

Comment (geoffw0) [reply rust/ql/lib/codeql/files/FileSystem.qll]: Done.

Though I am a bit worried the attached comments issue is going to turn out to affect things other than top-level items, I'll be happier when the issue with locations itself is fixed.

**Comment by geoffw0** [comment]
> Though I am a bit worried the attached comments issue is going to turn out to affect things other than top-level items, I'll be happier when the issue with locations itself is fixed.

I think I'm seeing this in the wild - things other than `Item`s with attached comments causing us to flag comment lines as code.  I wondering if the right answer for now is to only include end locations (as attached comments presumably always come _before_ the thing they're attached to)?  For single line items this won't make a difference, it'll avoid the attached comments issue, though we'll miss things like lines containing only `{`.

**Comment by geoffw0** [comment]
Done.  @aibaars I'd appreciate your thoughts on this.

**Comment by geoffw0** [comment]
With the latest changes we finally get an under-count rather than an over-count of lines of code compared to `cloc`.  This is what we should be getting.

**Comment by aibaars** [comment]
> I think I'm seeing this in the wild - things other than `Item`s with attached comments causing us to flag comment lines as code. I wondering if the right answer for now is to only include end locations (as attached comments presumably always come _before_ the thing they're attached to)? For single line items this won't make a difference, it'll avoid the attached comments issue, though we'll miss things like lines containing only `{`.

That would probably work. We could include the start line for `BlockExpr` and friends to avoid missing out on `{` .

I suspect that the following node kinds may all start with a comment:
```
 git grep 'HasDocComments for '
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Const {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Enum {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for ExternBlock {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for ExternCrate {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Fn {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Impl {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for MacroCall {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for MacroDef {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for MacroRules {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Module {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for RecordField {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for SourceFile {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Static {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Struct {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Trait {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for TraitAlias {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for TupleField {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for TypeAlias {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Union {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Use {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Variant {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Adt {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for AssocItem {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for ExternItem {}
crates/syntax/src/ast/generated/nodes.rs:impl ast::HasDocComments for Item {}
```

**Review by aibaars — APPROVED** [review]
Looks good to me. Perhaps include the start-line of BlockExpr and any other things that may start with a single `{` on a line.

**Comment by geoffw0** [comment]
I think I'd rather merge this now, further improvements can follow later if necessary.  When can I anticipate the locations themselves being fixed?  If it might be in the next few weeks, I'm inclined not to spend any more time iterating on workarounds.
